### PR TITLE
Expose value caster

### PR DESCRIPTION
**Pull request by henridf**
This commit does a minor refactoring of the cast-related code in package
expr in order to expose the core casting operations such that they can
be used directly on values.

### CODE REVIEW

File: expr/cast.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	case "bstring":
+		return castToBstring
+	case "bytes":
+		return castToBytes
+	default:
+		return nil
+	}
+}
+
+func castToIntN(min, max int64, typ zng.Type) func(zng.Value) (zng.Value, error) {
```
Comment (nwt) [inline expr/cast.go:49]: Nit: `typ` feels more prominent than `min` and `max`, so I'd make this this `castToIntN(typ zng.Type, min, max int64)`.

File: expr/cast.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		return zng.Value{typ, zng.EncodeInt(v)}, nil
+	}
+}
+
+var castToInt8 = castToIntN(math.MinInt8, math.MaxInt8, zng.TypeInt8)
+var castToInt16 = castToIntN(math.MinInt16, math.MaxInt16, zng.TypeInt16)
+var castToInt32 = castToIntN(math.MinInt32, math.MaxInt32, zng.TypeInt32)
+var castToInt64 = castToIntN(0, 0, zng.TypeInt64)
+
+func castToUintN(max uint64, typ zng.Type) func(zng.Value) (zng.Value, error) {
```
Comment (nwt) [inline expr/cast.go:66]: Nit: `typ` feels more prominent than `min` and `max`, so I'd make this this `castToUintN(typ zng.Type, min, max int64)`.

File: expr/cast.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		// XXX GC
+		return zng.Value{typ, zng.EncodeInt(v)}, nil
+	}
+}
+
+var castToInt8 = castToIntN(math.MinInt8, math.MaxInt8, zng.TypeInt8)
+var castToInt16 = castToIntN(math.MinInt16, math.MaxInt16, zng.TypeInt16)
+var castToInt32 = castToIntN(math.MinInt32, math.MaxInt32, zng.TypeInt32)
+var castToInt64 = castToIntN(0, 0, zng.TypeInt64)
```
Comment (nwt) [inline expr/cast.go:64]: Nit: These are more interesting than `castToIntN`, so I'd put them before it.

File: expr/cast.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		// XXX GC
+		return zng.Value{typ, zng.EncodeUint(v)}, nil
+	}
+}
+
+var castToUint8 = castToUintN(math.MaxUint8, zng.TypeUint8)
+var castToUint16 = castToUintN(math.MaxUint16, zng.TypeUint16)
+var castToUint32 = castToUintN(math.MaxUint32, zng.TypeUint32)
+var castToUint64 = castToUintN(0, zng.TypeUint64)
```
Comment (nwt) [inline expr/cast.go:81]: Nit: These are more interesting than `castToUintN`, so I'd put them before it.

File: expr/cast.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+		// rather than a tzng encoding, e.g., for time, an iso date instead of
+		// ns int.  For now, this works for numbers and IPs.  We will fix in a
+		// subsequent PR (see issue #1603).
+		result := zv.Type.StringOf(zv.Bytes, zng.OutFormatUnescaped, false)
+		return zng.Value{typ, zng.EncodeString(result)}, nil
+	}
+}
+
+var castToString = castToStringy(zng.TypeString)
+var castToBstring = castToStringy(zng.TypeBstring)
```
Comment (nwt) [inline expr/cast.go:140]: Nit: Nit: These are more interesting than `castToStringy`, so I'd put them before it.

File: expr/eval.go [hunk 6]
```diff
@@ -715,179 +712,26 @@ func NewCast(expr Evaluator, typ string) (Evaluator, error) {
...
-	case "time":
-		return &TimeCast{expr}, nil
-	case "string":
-		return &StringCast{expr, zng.TypeString}, nil
-	case "bstring":
-		return &StringCast{expr, zng.TypeBstring}, nil
-	case "bytes":
-		return &BytesCast{expr}, nil
-	default:
+
```
Comment (nwt) [inline expr/eval.go:715]:
```suggestion
```

File: expr/eval.go [hunk 7]
```diff
@@ -715,179 +712,26 @@ func NewCast(expr Evaluator, typ string) (Evaluator, error) {
...
-	// rather than a tzng encoding, e.g., for time, an iso date instead of
-	// ns int.  For now, this works for numbers and IPs.  We will fix in a
-	// subsequent PR (see issue #1603).
-	result := zv.Type.StringOf(zv.Bytes, zng.OutFormatUnescaped, false)
-	return zng.Value{s.typ, zng.EncodeString(result)}, nil
-}
-
-type BytesCast struct {
-	expr Evaluator
+type EvalCast struct {
```
Comment (nwt) [inline expr/eval.go:724]: Not sure if you're planning to use this outside package `eval`, but if not, maybe rename to `evalCast`.

**Review by nwt — APPROVED** [review]

File: expr/cast.go (hunk 1 above)
Comment (henridf) [reply expr/cast.go:49]: 👍

File: expr/cast.go (hunk 3 above)
Comment (henridf) [reply expr/cast.go:64]: Ha. I'd put them after due to some recollection of team preference towards putting things in order of use. 

(I don't mind either way and have adjusted.)

File: expr/eval.go (hunk 7 above)
Comment (henridf) [reply expr/eval.go:724]: 👍

**Pull request merged**